### PR TITLE
[FLINK-17084][table] Implement input type inference and output type strategy for ARRAY/ROW/MAP built-in functions

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

The type strategies are implemented using the new type stack. It also
removes those three planner expressions from the Blink planner as they
are no longer needed. Equivalent expressions in flink planner are not
yet removed as they are needed for converting to RexNodes.

It disables the FunctionDefinitionConvertRule for built-in functions as
the BridgingSqlFunction call generator supports only user defined functions so far.

## Verifying this change

Added tests in `InputTypeStrategiesTest` and `TypeStrategiesTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no /** don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java [hunk 1]
```diff
@@ -553,5 +609,10 @@ TestSpec expectErrorMessage(String expectedErrorMessage) {
 			this.expectedErrorMessage = expectedErrorMessage;
 			return this;
 		}
+
+		@Override
+		public String toString() {
+			return description != null ? description : "";
```
Comment (twalthr) [inline flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/InputTypeStrategiesTest.java:615]: nit: how about we just use the class name of input strategy by default

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/validation/ArrayTypeValidationTest.scala [hunk 2]
```diff
@@ -26,11 +26,6 @@ import org.junit.Test
 
 class ArrayTypeValidationTest extends ArrayTypeTestBase {
 
-  @Test(expected = classOf[ValidationException])
-  def testImplicitTypeCastTableApi(): Unit = {
```
Comment (twalthr) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/expressions/validation/ArrayTypeValidationTest.scala:30]: Can we add a test to `org.apache.flink.table.planner.expressions.ArrayTypeTest` for this if it works now.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/expressions/validation/MapTypeValidationTest.scala [hunk 3]
```diff
@@ -45,11 +45,6 @@ class MapTypeValidationTest extends MapTypeTestBase {
     testAllApis("FAIL", "map()", "MAP[]", "FAIL")
   }
 
-  @Test(expected = classOf[ValidationException])
-  def testUnsupportedMapImplicitTypeCastTableApi(): Unit = {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/expressions/validation/MapTypeValidationTest.scala:49]: Add a test to `org.apache.flink.table.planner.expressions.MapTypeTest`?

**Comment by dawidwys** [comment]
I added the tests you asked for. I found out that it actually does not work for sql (works well for table api in this PR). I created a jira issue to track it: https://issues.apache.org/jira/browse/FLINK-17144

**Pull request merged**